pkg/koordlet/resourceexecutor/resctrl.go

```diff
@@ -0,0 +1,160 @@
+/*
+Copyright 2022 The Koordinator Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resourceexecutor
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"k8s.io/klog/v2"
+
+	"github.com/koordinator-sh/koordinator/pkg/koordlet/util/system"
+)
+
+const ErrResctrlDir = "resctrl path or file not exist"
+const CacheIdIndex = 2
+
+// NewResctrlReader: lazy resctrl reader, just check vendor to generate specific reader
+func NewResctrlReader() ResctrlReader {
+	// Support two main platforms; other platforms need to add their implementation of the resctrl interface.
+	if vendorId, err := system.GetVendorIDByCPUInfo(system.GetCPUInfoPath()); err != nil {
+		klog.V(0).ErrorS(err, "get cpu vendor error, stop start resctrl collector")
+		return &fakeReader{}
+	} else {
+		switch vendorId {
+		case system.INTEL_VENDOR_ID:
+			return NewResctrlRDTReader()
+		case system.AMD_VENDOR_ID:
+			return NewResctrlQoSReader()
+		default:
+			klog.V(0).ErrorS(err, "unsupported cpu vendor")
+		}
+	}
+	return &fakeReader{}
+}
+
+type CacheId int
+
+// parent for resctrl is like: `BE`, `LS`
+type ResctrlReader interface {
+	ReadResctrlL3Stat(parent string) (map[CacheId]uint64, error)
+	ReadResctrlMBStat(parent string) (map[CacheId]system.MBStatData, error)
+}
+
+type ResctrlBaseReader struct {
+}
+
+type ResctrlRDTReader struct {
+	ResctrlBaseReader
+}
+type ResctrlAMDReader struct {
+	ResctrlBaseReader
+}
+
+type fakeReader struct {
+	ResctrlBaseReader
+}
+
+func (rr *fakeReader) ReadResctrlL3Stat(parent string) (map[CacheId]uint64, error) {
+	return nil, errors.New("unsupported platform")
+}
+
+func (rr *fakeReader) ReadResctrlMBStat(parent string) (map[CacheId]system.MBStatData, error) {
+	return nil, errors.New("unsupported platform")
+}
+
+func NewResctrlRDTReader() ResctrlReader {
+	return &ResctrlRDTReader{}
+}
+
+func NewResctrlQoSReader() ResctrlReader {
+	return &ResctrlAMDReader{}
+}
+
+func (rr *ResctrlBaseReader) ReadResctrlL3Stat(parent string) (map[CacheId]uint64, error) {
```

Comment:
more comments about ReadResctrlL3Stat/ReadResctrlMBStat can make the reader easy understand